app/plugin.go

```diff
@@ -624,3 +604,92 @@ func (a *App) notifyPluginEnabled(manifest *model.Manifest) error {
+	}
+	if len(fileStorePaths) == 0 {
+		mlog.Info("Found no files in plugins file store")
+		return nil, nil
+	}
+	pluginSignaturePathMap := make(map[string]*pluginSignaturePaths)
+	for _, path := range fileStorePaths {
+		if strings.HasSuffix(path, ".tar.gz") {
+			id := strings.TrimSuffix(filepath.Base(path), ".tar.gz")
+			helper := &pluginSignaturePaths{
+				path:           path,
+				signaturePaths: []string{},
+			}
+			pluginSignaturePathMap[id] = helper
+		}
+	}
+	for _, path := range fileStorePaths {
+		if strings.HasSuffix(path, ".sig") {
+			id := strings.TrimSuffix(filepath.Base(path), ".sig")
+			index := strings.LastIndex(id, ".")
+			id = id[:index]
+			if val, ok := pluginSignaturePathMap[id]; !ok {
+				mlog.Error("Unknown signature in the filestore.", mlog.String("path", path))
+			} else {
+				val.signaturePaths = append(val.signaturePaths, path)
+			}
+		}
+	}
+	return pluginSignaturePathMap, nil
+}
+
+func (a *App) verifySignature(plugin filesstore.ReadCloseSeeker, signaturePaths []string) *model.AppError {
+	for _, signaturePath := range signaturePaths {
+		var sigReader filesstore.ReadCloseSeeker
+		sigReader, appErr := a.FileReader(signaturePath)
+		if appErr != nil {
+			mlog.Debug("Can't read signature file.", mlog.String("path", signaturePath), mlog.Err(appErr))
+			continue
+		}
+		plugin.Seek(0, 0)
+		if err := a.VerifyPlugin(plugin, sigReader); err != nil {
+			mlog.Debug("Failed to verify plugin signature in the file store.", mlog.String("signature", signaturePath), mlog.Err(err))
+			continue
+		}
+		return nil
+	}
+	return model.NewAppError("SyncPlugins", "app.plugin.sync_plugins.verify_signature.app_error", nil, "", http.StatusInternalServerError)
+
+}
+
+func (a *App) installPluginsWithSignaturePaths(plugins []*pluginSignaturePaths) {
+	for _, plugin := range plugins {
+		if _, err := a.installPluginWithSignaturePath(plugin); err != nil {
+			mlog.Error("Failed to install plugin.", mlog.String("bundle", plugin.path), mlog.Err(err))
+		}
+	}
+}
+
+func (a *App) installPluginWithSignaturePath(plugin *pluginSignaturePaths) (*model.Manifest, *model.AppError) {
+	var reader filesstore.ReadCloseSeeker
+	reader, appErr := a.FileReader(plugin.path)
+	if appErr != nil {
+		return nil, model.NewAppError("installPluginWithSignaturePath", "app.plugin.install_plugin_with_sig.open.app_error", nil, "", http.StatusInternalServerError)
+
+	}
+	defer reader.Close()
+	if *a.Config().PluginSettings.RequirePluginSignature {
+		if appErr = a.verifySignature(reader, plugin.signaturePaths); appErr != nil {
+			return nil, model.NewAppError("installPluginWithSignaturePath", "app.plugin.install_plugin_with_sig.verify.app_error", nil, "", http.StatusInternalServerError)
+		}
+	}
+	reader.Seek(0, 0)
+	mlog.Info("Syncing plugin from file store", mlog.String("bundle", plugin.path))
+	manifest, err := a.installPluginLocally(reader, true)
```

Comment:
i feel like all this verification logic belongs inside `installPluginLocally`. With all these different ways of installing a plugin i worry we might end up creating multiple code paths to installing a plugin. What if we add an optional parameter to `installPluginLocally` that takes `signatures []io.Reader` or something?